### ISSUE TITLE
bug:  MemGPT, Text-Generation-WebUI, and HF models -and Transformers- trimming prompt

### ISSUE THREAD

**Issue by madgrizzle** [letta-ai/letta#1398]
-edit- found out more about the issue and added another comment as it affects transformers as well.

**Describe the bug**
This issue seems to be related to using HF model loaders.  I don't know what other servers use `llamacpp_HF` or `ExLlamav2_HF` model loaders, but text-generation-webui (TGW) does and the bug readily shows up there.  It doesn't show up in non-HF model loaders that I can find because they appear to not call the same routines that result in the bug.

I've noticed an oddity in TGW with the --verbose enabled and see that the beginning of the prompt is being trimmed.  The difference is related to how MemGPT calculates the number of tokens of a prompt vs. what the server calculates it.  MemGPT uses tiktoken with a default "gpt-4" model whereas `llamacpp_HF` and `ExLlamav2_HF`  uses tokenizer.Tokenizer and uses the tokenizer.json file. 

For a given prompt, tiktoken encoding will result in fewer tokens than HF's Tokenizer class.  In MemGPT, the max_token value in the api request is calculated to be the context_size - prompt_tokens.  This leads to the parameters being sent for a prompt of, for example, 7361 tokens to have a max_tokens of 831 for a context length of 8192.

In TGW, the same prompt (and this is using @starsnatched MemGPT_DPO_MoE_test model that I quantized with exllama) gets 8360 tokens returned when encoded.  With a max_token_count of 831, the total size is calculated as 8360+831=9191. 
 With a context_size of 8192, then the prompt gets trimmed from the beginning by 9191-8192=999 tokens:

```
    if truncation_length is not None:
        input_ids = input_ids[:, -truncation_length:]
```

Because memgpt sends the truncation_length and automatically maxes out the max_new_tokens in the parameters sent to TGW (therefore overriding any set parameters in the webui) the prompt is always calculated as being too long and getting trimmed. 

I verified this is the case with both `llamacpp_HF` and `ExLlamav2_HF`  loaders in TWG.  If I comment out the line in memgpt that adds the context_length to the request and manually set the context_length in TGW to be something a bit bigger than 8192 (and set alpha to 1.75) to give some cushion, then it works, but this takes up more GPU memory.  I don't know the advantage/disadvantages of using the llamacpp vs llamacpp_HF, but the HF version comes up as the default in TGW for exl2 files so it might be a source of issue for some people.

**Comment by madgrizzle** [comment]
Well, that doesn't really work either because model seems to fall apart after 8K context limit is reached (starsnatched's model card says it works up to 8K).

**Comment by madgrizzle** [comment]
This problem appears to affect transformers as the model loader as well.  I loaded @starsnatched 's latest model in half-precision and loaded it using the default transformers model loader.  I have modified the TGW code to report the length of the prompt that it calculates and it too is reporting a value different than what MemGPT sends it.  The effect, like with HF model loaders, is to trim the front of the system prompt to accommodate the max_token_count.

It's easy to test in TGW.  Just start it with --verbose argument and look at the prompt that gets outputted to the screen and you'll start to see the beginning of the prompt being trimmed.
